internal/tofu/node_provider.go

```diff
@@ -26,33 +27,80 @@ var (
 )
 
 // GraphNodeExecutable
-func (n *NodeApplyableProvider) Execute(ctx EvalContext, op walkOperation) (diags tfdiags.Diagnostics) {
-	_, err := ctx.InitProvider(n.Addr)
-	diags = diags.Append(err)
+func (n *NodeApplyableProvider) Execute(ctx EvalContext, op walkOperation) tfdiags.Diagnostics {
+	instances, diags := n.initInstances(ctx, op)
+
+	for key, provider := range instances {
+		diags = diags.Append(n.executeInstance(ctx, op, key, provider))
+	}
+
+	return diags
+}
+func (n *NodeApplyableProvider) initInstances(ctx EvalContext, op walkOperation) (map[addrs.InstanceKey]providers.Interface, tfdiags.Diagnostics) {
+	var diags tfdiags.Diagnostics
+
+	var initKeys []addrs.InstanceKey
+	if n.Config == nil || n.Config.Instances == nil || op == walkValidate {
+		initKeys = append(initKeys, addrs.NoKey)
+	} else {
+		// Instances are set AND we are not validating
+		for key := range n.Config.Instances {
+			initKeys = append(initKeys, key)
+		}
+	}
+
+	// init -> config (different due to validate skipping most for_each logic)
+	instanceKeys := make(map[addrs.InstanceKey]addrs.InstanceKey)
+	if n.Config == nil || n.Config.Instances == nil {
+		instanceKeys[addrs.NoKey] = addrs.NoKey
+	} else if op == walkValidate {
+		// Instances are set AND we are validating
+		for key := range n.Config.Instances {
+			instanceKeys[addrs.NoKey] = key
+		}
```

Comment:
nit: from the first sight, it looks like a typo. Let's make it explicit that we just want to get any existing key:

```suggestion
		for key := range n.Config.Instances {
			instanceKeys[addrs.NoKey] = key
			break
		}
```

Reply:
This is actually a bug

Reply:
Fixed in 7fd09e5d2a